libs/framework/src/celix_bundle_cache.c

```diff
@@ -71,124 +137,223 @@ celix_status_t celix_bundleCache_create(celix_framework_t* fw, celix_bundle_cach
+    celix_status_t status = CELIX_SUCCESS;
+    char delims[] = " ";
+    char *savePtr = NULL;
+    char zipFileListBuffer[CELIX_DEFAULT_STRING_CREATE_BUFFER_SIZE];
+    char* zipFileList = celix_utils_writeOrCreateString(zipFileListBuffer, sizeof(zipFileListBuffer), "%s", list);
+    if (zipFileList) {
+        char *location = strtok_r(zipFileList, delims, &savePtr);
+        while (location != NULL) {
+            bundle_archive_t* archive = NULL;
+            status = celix_bundleCache_createArchive(fw, (*bndId)++, location, &archive);
+            if (status != CELIX_SUCCESS) {
+                fw_logCode(fw->logger, CELIX_LOG_LEVEL_ERROR, CELIX_BUNDLE_EXCEPTION, "Cannot create bundle archive for %s", location);
+                break;
+            } else {
+                celix_log_level_e lvl = logProgress ? CELIX_LOG_LEVEL_INFO : CELIX_LOG_LEVEL_DEBUG;
+                fw_log(fw->logger, lvl, "Created bundle cache '%s' for bundle archive %s (bndId=%li).",
+                       celix_bundleArchive_getCurrentRevisionRoot(archive),
+                       celix_bundleArchive_getSymbolicName(archive), celix_bundleArchive_getId(archive));
+                bundleArchive_destroy(archive);
+            }
+            location = strtok_r(NULL, delims, &savePtr);
+        }
+    } else {
+        status = CELIX_ENOMEM;
+        fw_logCode(fw->logger, CELIX_LOG_LEVEL_ERROR, status, "Failed to create zip file list.");
+    }
+    celix_utils_freeStringIfNeeded(zipFileListBuffer, zipFileList);
+    return status;
 }
 
-celix_status_t celix_bundleCache_createArchive(celix_framework_t* fw, long id, const char * location, const char *inputFile, bundle_archive_pt *archive) {
-	celix_status_t status = CELIX_SUCCESS;
-	char archiveRoot[512];
+//using tmp cache and that bundle cache is not deleted during bundle cache destroy.
+celix_status_t celix_bundleCache_createBundleArchivesCache(celix_framework_t *fw, bool logProgress) {
+    celix_status_t status = CELIX_SUCCESS;
 
-	if (fw && location) {
-		snprintf(archiveRoot, sizeof(archiveRoot), "%s/bundle%ld",  fw->cache->cacheDir, id);
-		status = bundleArchive_create(fw, archiveRoot, id, location, inputFile, archive);
-	}
+    const char* const cosgiKeys[] = {"cosgi.auto.start.0","cosgi.auto.start.1","cosgi.auto.start.2","cosgi.auto.start.3","cosgi.auto.start.4","cosgi.auto.start.5","cosgi.auto.start.6", NULL};
+    const char* const celixKeys[] = {CELIX_AUTO_START_0, CELIX_AUTO_START_1, CELIX_AUTO_START_2, CELIX_AUTO_START_3, CELIX_AUTO_START_4, CELIX_AUTO_START_5, CELIX_AUTO_START_6, NULL};
+    CELIX_BUILD_ASSERT(sizeof(*cosgiKeys) == sizeof(*celixKeys));
+    long bndId = CELIX_FRAMEWORK_BUNDLE_ID+1; //note cleaning cache, so starting bundle id at 1
 
-	framework_logIfError(fw->logger, status, NULL, "Failed to create archive");
+    const char* errorStr = NULL;
+    status = celix_utils_deleteDirectory(fw->cache->cacheDir, &errorStr);
```

Comment:
A malicious directory may cause infinite recursion, see #489.

Reply:
Good catch and indeed something we should fix. Seeing that you created a issue for this, I assume that we will do this in a separate PR.

Reply:
Yes, I'm dealing with the symbolic link. But eliminating recursion should be done in a separate PR.